### PR TITLE
Add cruft pack support

### PR DESCRIPTION
**Pull request by user1736**

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -1,19 +1,31 @@
...
+import { createPackedObjectIndex, CruftPackMode } from './packed-object-index.js';
 import { createFilesMethods } from './files-list.js';
 import { createCommitMethods } from './commits.js';
 import { createStatMethod } from './stat.js';
 
 export * from './types.js';
 export * from './parse-object.js';
-export async function createGitReader(gitdir: string) {
+
+export interface GitReaderParams {
```
Comment (lahmatiy) [inline src/index.ts]: It should be moved to `types.ts`, and `Params` -> `Options`

File: src/index.ts [hunk 2]
```diff
@@ -1,19 +1,31 @@
...
+    /**
+     * Controls the inclusion of cruft packs in packed objects procession.
+     * @see {@link https://git-scm.com/docs/cruft-packs} for more info about cruft packs.
+     * @default 'include'
+     */
+    cruftPacks?: CruftPackMode;
+}
+
+export async function createGitReader(gitdir: string, params?: GitReaderParams) {
```
Comment (lahmatiy) [inline src/index.ts]: `params` -> `options`, since `params` is already a parameter

File: src/packed-object-index.ts [hunk 3]
```diff
@@ -9,7 +9,14 @@ import {
 } from './types.js';
 import { objectsStatFromTypes, sumObjectsStat } from './utils/stat.js';
 
-export async function createPackedObjectIndex(gitdir: string) {
+export type CruftPackMode = 'include' | 'exclude' | 'cruft-only';
```
Comment (lahmatiy) [inline src/packed-object-index.ts]: Should be moved to `types.ts`. And since you add a descriptive comments, it would be nice to describe the meaning of each value.
I think that `true` should be an alias for `"include"` and `false` for `"exclude"`.  Nit, probably `"cruft-only"` might be renamed into `"only"`, i.e. `cruftPacks: "cruft-only"` vs. `cruftPacks: "only"`

File: src/packed-object-index.ts [hunk 4]
```diff
@@ -43,8 +50,24 @@ export async function createPackedObjectIndex(gitdir: string) {
     }
 
     const packdir = pathJoin(gitdir, 'objects/pack');
-    const packFilenames = (await fsPromises.readdir(packdir))
-        .filter((filename) => filename.endsWith('.pack'))
+    const packdirFiles = await fsPromises.readdir(packdir);
+    const cruftPackNames =
```
Comment (lahmatiy) [inline src/packed-object-index.ts]: `Names` -> `Filenames`

File: src/packed-object-index.ts [hunk 5]
```diff
@@ -43,8 +50,24 @@ export async function createPackedObjectIndex(gitdir: string) {
...
+        cruftPackMode !== 'include'
+            ? packdirFiles.filter((f) => f.endsWith('.mtimes')).map((f) => f.slice(0, -7))
+            : [];
+
+    const packFilenames = packdirFiles
+        .filter((filename) => {
+            const isPackFile = filename.endsWith('.pack');
+            if (cruftPackMode === 'include' || !isPackFile) {
```
Comment (lahmatiy) [inline src/packed-object-index.ts]: Let's simplify this part. Currently, it's pretty hard to follow what's will be returned and when.

```js
// accept filenames with a .pack extension only
if (!filename.endsWith('.pack')) {
    return false;
}

// use all pack disregarding of its type, when cruft pack mode is inclusive
if (cruftPackMode === 'include') {
    return true;
}
```

File: src/packed-object-index.ts [hunk 6]
```diff
@@ -43,8 +50,24 @@ export async function createPackedObjectIndex(gitdir: string) {
...
+                return isPackFile;
+            }
+
+            const fnWithoutExt = filename.slice(0, -5);
```
Comment (lahmatiy) [inline src/packed-object-index.ts]: The same as above, it is not error-proof:

```js
const fnWithoutExt = path.basename(filename, '.pack');
```

File: src/packed-object-index.ts [hunk 7]
```diff
@@ -43,8 +50,24 @@ export async function createPackedObjectIndex(gitdir: string) {
     }
 
     const packdir = pathJoin(gitdir, 'objects/pack');
-    const packFilenames = (await fsPromises.readdir(packdir))
-        .filter((filename) => filename.endsWith('.pack'))
+    const packdirFiles = await fsPromises.readdir(packdir);
+    const cruftPackNames =
+        cruftPackMode !== 'include'
+            ? packdirFiles.filter((f) => f.endsWith('.mtimes')).map((f) => f.slice(0, -7))
```
Comment (lahmatiy) [inline src/packed-object-index.ts]: The `.slice()` operation with a magic number `-7 ` is not error-proof. Much better to be clear:

```js
path.basename(filename, '.mtimes');
```

File: src/packed-object-index.ts [hunk 8]
```diff
@@ -9,7 +9,14 @@ import {
...
 
-export async function createPackedObjectIndex(gitdir: string) {
+export type CruftPackMode = 'include' | 'exclude' | 'cruft-only';
+
+/**
+ * Creates an index object that provides information about git objects stored within pack files
+ * @param gitdir absolute path to .git dir
+ * @param cruftPackMode controls the inclusion of cruft packs in packed objects procession. @see {@link https://git-scm.com/docs/cruft-packs} for more info about cruft packs.
+ */
+export async function createPackedObjectIndex(gitdir: string, cruftPackMode: CruftPackMode) {
```
Comment (lahmatiy) [inline src/packed-object-index.ts]: I believe, there will be more options (for instance about cache control, maybe smth else). So better pass an options as is here.

File: src/packed-object-index.ts [hunk 9]
```diff
@@ -9,7 +9,14 @@ import {
 } from './types.js';
 import { objectsStatFromTypes, sumObjectsStat } from './utils/stat.js';
 
-export async function createPackedObjectIndex(gitdir: string) {
+export type CruftPackMode = 'include' | 'exclude' | 'cruft-only';
+
+/**
+ * Creates an index object that provides information about git objects stored within pack files
+ * @param gitdir absolute path to .git dir
+ * @param cruftPackMode controls the inclusion of cruft packs in packed objects procession. @see {@link https://git-scm.com/docs/cruft-packs} for more info about cruft packs.
```
Comment (lahmatiy) [inline src/packed-object-index.ts]: DRY. The option is already described on `GitReaderOptions`

File: src/index.ts [hunk 10]
```diff
@@ -1,19 +1,31 @@
...
+    const paramsWithDefaults = Object.assign({ cruftPacks: 'include' }, params);
```
Comment (lahmatiy) [inline src/index.ts]: That's a good practice to have an options normalisation function, which may throws when a wrong value for an option or unknown option is passed.

File: fixtures/cruft/_git/COMMIT_EDITMSG [hunk 11]
```diff
@@ -0,0 +1 @@
+Test commit
```
Comment (lahmatiy) [inline fixtures/cruft/_git/COMMIT_EDITMSG]: I think we can remove all unnecessary files for this fixture like `logs/*`, `object/*` (except packs), `refs/tags/*` etc

**Review by lahmatiy — CHANGES_REQUESTED** [review]

File: src/index.ts (hunk 10 above)
Comment (user1736) [reply src/index.ts]: An example of the function format would be a huge help here since I can imagine quite a few ways of doing it.

File: src/index.ts (hunk 1 above)
Comment (user1736) [reply src/index.ts]: why though? I don't see much value in moving this interface since here's the only place where it's used

File: fixtures/cruft/_git/COMMIT_EDITMSG (hunk 11 above)
Comment (user1736) [reply fixtures/cruft/_git/COMMIT_EDITMSG]: btw, if I remove `tags`/`remotes` directories out the tool explodes. I put `.gitkeep` there for now, to preserve the directories on checkout.

Comment (user1736) [reply fixtures/cruft/_git/COMMIT_EDITMSG]: okay, `.gitkeep` is also not a great idea here, gonna look up a better compromise

Comment (user1736) [reply fixtures/cruft/_git/COMMIT_EDITMSG]: added checks whether dir exists before reading refs/tags in https://github.com/discoveryjs/scan-git/pull/5/commits/4265b6e3529671dff79c35972e5842c60f05b38d

File: src/index.ts (hunk 10 above)
Comment (user1736) [reply src/index.ts]: got the example in the chat

File: src/index.ts (hunk 1 above)
Comment (user1736) [reply src/index.ts]: discussed, moved the interface

File: src/packed-object-index.ts [hunk 12]
```diff
@@ -1,15 +1,33 @@
...
+     * Controls the inclusion of cruft packs in packed objects procession.
+     */
+    cruftPacks: CruftPackMode;
+}
+
+/**
+ * Creates an index object that provides information about git objects stored within pack files
+ * @param options
+ */
+export async function createPackedObjectIndex(options: PackedObjectIndexOptions) {
```
Comment (lahmatiy) [inline src/packed-object-index.ts]: If you change the signature to the following, you will avoid unnecessary type declarations (less code):
```suggestion
export async function createPackedObjectIndex(gitdir: string, options: GitReaderOptions) {
```

File: src/resolve-ref.ts [hunk 13]
```diff
@@ -65,13 +65,16 @@ export async function createRefIndex(gitdir: string) {
...
     const listRemotes = async () => {
         const remotes = [];
-        const entries = await fsPromises.readdir(pathJoin(gitdir, 'refs', 'remotes'), {
-            withFileTypes: true
-        });
+        const remotesDir = pathJoin(gitdir, 'refs', 'remotes');
+        if (existsSync(remotesDir)) {
+            const entries = await fsPromises.readdir(pathJoin(gitdir, 'refs', 'remotes'), {
+                withFileTypes: true
+            });
```
Comment (lahmatiy) [inline src/resolve-ref.ts]: Maybe in case of `!existsSync(remotesDir)` just use an empty array?

```suggestion
        const entries = existsSync(remotesDir)
            ? await fsPromises.readdir(pathJoin(gitdir, 'refs', 'remotes'), {
                   withFileTypes: true
               })
            : [];
```

File: src/resolve-ref.ts [hunk 14]
```diff
@@ -84,32 +87,37 @@ export async function createRefIndex(gitdir: string) {
         let refs = listRefsCache.get(prefix);
 
         if (refs === undefined) {
-            const packedRefs = await readPackedRefs(gitdir);
```
Comment (lahmatiy) [inline src/resolve-ref.ts:87]: These changes are not needed, since `readPackedRefs()` checks a path existence and returns an empty map when it doesn't exist.

File: src/types.ts [hunk 15]
```diff
@@ -112,3 +112,20 @@ type TreeDiffModified = {
 };
 export type TreeDiffEntry = TreeDiffAdded | TreeDiffRemoved | TreeDiffModified;
 export type TreeDiff = TreeDiffEntry[];
+
+export type CruftPackMode =
+    /** Processes all packs */
+    | 'include'
```
Comment (lahmatiy) [inline src/types.ts:118]: You didn't accept my suggestions to add `true` / `false` aliases for `"include"` / `"exclude"`, or missed it?

File: test/stat.ts [hunk 16]
```diff
@@ -2,16 +2,42 @@ import assert from 'assert';
...
     describe('stat()', () => {
         it('without params', async () => {
+            const repo = await fixtures.base.repo();
             const actual = await repo.stat();
 
-            assert.deepStrictEqual(actual, data.stat);
+            assert.deepStrictEqual(actual, fixtures.base.data.stat);
+        });
+
+        it('counts objects in cruft packs', async () => {
```
Comment (lahmatiy) [inline test/stat.ts:13]: A case for `fixtures.cruft.repo()` with no options is needed here to test a default value.

**Review by lahmatiy — CHANGES_REQUESTED** [review]

File: src/packed-object-index.ts (hunk 12 above)
Comment (user1736) [reply src/packed-object-index.ts]: yup, I considered this, however `PackedObjectIndexOptions` and `GitReaderOptions` are somewhat different. The first one is a subset of the later one.

File: src/resolve-ref.ts (hunk 14 above)
Comment (user1736) [reply src/resolve-ref.ts:87]: I will double-check it, but it did explode on the fixture

Comment (lahmatiy) [reply src/resolve-ref.ts:87]: It shouldn't explode for sure since `packed-refs` is optional feature

Comment (user1736) [reply src/resolve-ref.ts:87]: alright, so the problem boils down to the following: now `listRefs` is used inside `listBranches` and `listTags`

`listTags` would explode if `tags` directory is missing, which is exactly the case for current fixture (it's a pain to preserve empty folder in git structure)

Speaking of the `listBranches` , it's available as a part of the `gitReader` API, meaning that consumers could pass any remote name as an argument which would result in an undescriptive error

File: src/packed-object-index.ts (hunk 12 above)
Comment (lahmatiy) [reply src/packed-object-index.ts]: There is no need for a specialized options for internal parts. It's also simplify things and number of changes when a new options will be added

File: test/stat.ts [hunk 17]
```diff
@@ -2,16 +2,52 @@ import assert from 'assert';
...
+            );
+        });
+
+        it('counts all packed objects, when cruftPacks option is not specified', async () => {
+            const repo = await fixtures.cruft.repo();
+            const stat = await repo.stat();
+
+            assert.deepStrictEqual(
+                stat.objects.packed.objects.count,
+                fixtures.cruft.data.totalPackedObjects
```
Comment (lahmatiy) [inline test/stat.ts:49]: Wrong default, should be `cruftPacks: "exclude"` (we don't need for cruft objects by default since that's marked for deletion objects)

**Review by lahmatiy — APPROVED** [review]

**Comment by lahmatiy** [comment]
Thank you!
